events/ratelimiting/coalescing.go

```diff
@@ -0,0 +1,251 @@
+	defer cancel()
+
+	for {
+		// If the timer doesn't exist yet, we're waiting for the first event (which
+		// will fire immediately when received).
+		var timerCh <-chan time.Time
+		c.lock.RLock()
+		if c.hasTimer.Load() {
+			timerCh = c.timer.C()
+		}
+		c.lock.RUnlock()
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-c.closeCh:
+			cancel()
+			return nil
+
+		case <-c.inputCh:
+			c.handleInputCh(ctx, ch)
+
+		case <-timerCh:
+			c.handleTimerFired(ctx, ch)
+		}
+	}
+}
+
+func (c *coalescing) handleInputCh(ctx context.Context, ch chan<- struct{}) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	switch {
+	case !c.hasTimer.Load():
+		// We don't have a timer yet, so this is the first event that has fired. We
+		// fire the event immediately, and set the timer to fire again after the
+		// initial delay.
+		c.timer = c.clock.NewTimer(c.initialDelay)
+		c.hasTimer.Store(true)
+		c.fireEvent(ctx, ch)
+
+	default:
+		// If maxPendingEvents is set and we have reached it then fire the event
+		// immediately.
+		if c.maxPendingEvents != nil && c.pendingEvents >= *c.maxPendingEvents {
+			c.fireEvent(ctx, ch)
+			return
+		}
+
+		if !c.timer.Stop() {
+			<-c.timer.C()
+		}
+
+		// Setup backoff. Backoff is exponential. If initial is 500ms and max is
+		// 5s, the backoff will follow:
+		// 500ms, 1s, 2s, 4s, 5s, 5s, 5s, ...
+		if c.currentDur < c.maxDelay {
+			c.backoffFactor *= 2
+			c.currentDur = time.Duration(float64(c.initialDelay) * float64(c.backoffFactor))
+			if c.currentDur > c.maxDelay {
+				c.currentDur = c.maxDelay
+			}
+		}
+
+		c.timer.Reset(c.currentDur)
+	}
+}
+
+func (c *coalescing) handleTimerFired(ctx context.Context, ch chan<- struct{}) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.fireEvent(ctx, ch)
+	c.reset()
+}
+
+func (c *coalescing) fireEvent(ctx context.Context, ch chan<- struct{}) {
+	// Important to only send on the channel if there are pending events,
+	// otherwise we will double send an event, for example if only a single event
+	// was sent and then the rate limiting window expired with no new events.
+	if c.pendingEvents > 0 {
+		c.pendingEvents = 0
+		c.wg.Add(1)
+		go func() {
+			defer c.wg.Done()
+			select {
+			case ch <- struct{}{}:
+			case <-ctx.Done():
+			}
+		}()
+	}
+}
+
+func (c *coalescing) reset() {
+	if !c.timer.Stop() {
+		select {
+		case <-c.timer.C():
+		default:
+		}
+	}
+
+	c.pendingEvents = 0
+	c.currentDur = c.initialDelay
+	c.backoffFactor = 1
+	c.hasTimer.Store(false)
+	c.timer = nil
+}
+
+func (c *coalescing) Add() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.pendingEvents++
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		select {
+		case c.inputCh <- struct{}{}:
+		case <-c.closeCh:
+		}
+	}()
+}
+
+func (c *coalescing) Close() {
+	defer func() {
+		// Prevent wg race condition on Close and Run.
+		c.lock.Lock()
+		c.wg.Wait()
+		c.lock.Unlock()
+	}()
+	if c.closed.CompareAndSwap(false, true) {
+		close(c.closeCh)
+	}
+}
+
+var _ RateLimiter = new(coalescing)
```

Comment:
Using `new` causes the allocation of the object which stays around forever

```suggestion
var _ RateLimiter = (*coalescing)(nil)
```

Reply:
Done